jplag/src/main/java/de/jplag/clustering/Preprocessor.java

```diff
@@ -0,0 +1,78 @@
+package de.jplag.clustering;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.DoubleStream;
+
+import org.apache.commons.math3.linear.Array2DRowRealMatrix;
+import org.apache.commons.math3.linear.RealMatrix;
+
+import de.jplag.clustering.algorithm.ClusteringAlgorithm;
+
+/**
+ * Interface for classes that process similarity matrices before any clustering. Classes implementing this interface
+ * must ensure that they do not produce zero rows/columns inside the similarity matrix. They must also be able to
+ * calculate the original indices of rows/columns after use through their {@link Preprocessor#postProcessResult} method.
+ */
+public interface Preprocessor {
+    double[][] preprocessSimilarities(double[][] similarityMatrix);
+
+    Collection<Collection<Integer>> postProcessResult(Collection<Collection<Integer>> result);
+
```

Comment:
Ist the naming postProcessResult intentional or would postPreprocessResult be a better fitting name?

Reply:
Even more: Is a collection<collection<Integer>> a good type here?

Reply:
It's like this: Some preprocessing might introduce rows or columns that are zero and a problem for spectral clustering. So pre-processors must ensure that there are no zero rows/cols (remove them).
But this is changes the indices in the results from the clustering algorithm.
So the pre-processor is also expected to do a post-processing on the results so that the indices match with the indices of the original similarity matrix.

I've changed the interface to only map back single indices not collections of collections but that only moved stuff around.